### PR TITLE
PCVGuardian

### PR DESCRIPTION
**Pull request by ditchfieldcaleb**
PCV Guardian interface, implementation, and unit tests.

The PCV Guardian functions will (mostly) be only callable by the guardian-or-governor.

The Governor is the only role that will be able to add "safe" addresses that the guardian can choose to withdraw to.

The Guardian can call into the PCV Guardian (which itself will be a pcv controller and a guardian) and withdraw from any pcv deposit into any safe address.

### CODE REVIEW

File: contracts/pcv/PCVGuardian.sol [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import "./IPCVDeposit.sol";
+
+contract PCVGuardian is IPCVGuardian, CoreRef {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    // If an address is in this set, it's a safe address to withdraw to
+    EnumerableSet.AddressSet private safeAddresses;
+
+    constructor(
+        address _core
```
Comment (Joeysantoro) [inline contracts/pcv/PCVGuardian.sol:16]: add initial set of safeAddresses to constructor

File: contracts/pcv/PCVGuardian.sol [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    /// @notice returns all safe addresses
+    function getSafeAddresses() public view override returns (address[] memory) {
+        return safeAddresses.values();
+    }
+
+    // ---------- Governor-Only State-Changing API ----------
+
+    /// @notice governor-only method to set an address as "safe" to withdraw funds to
+    /// @param anAddress the address to set as safe
+    function setSafeAddress(address anAddress) external override onlyGovernor() {
```
Comment (Joeysantoro) [inline contracts/pcv/PCVGuardian.sol:38]: This should be onlyGovernorOrAdmin imo. Also add batch methods

File: contracts/pcv/IPCVGuardian.sol [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+/// @title IPCVGuardian
+/// @notice an interface for defining how the PCVGuardian functions
+/// @dev any implementation of this contract should be granted the roles of Guardian and PCVController in order to work correctly
+interface IPCVGuardian {
+    // ---------- Events ----------
+    event SafeAddressAdded(
+        address indexed anAddress
```
Comment (Joeysantoro) [inline contracts/pcv/IPCVGuardian.sol:10]: nit: safeAddress is better name

File: contracts/pcv/IPCVGuardian.sol [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    );
+
+    event SafeAddressRemoved(
+        address indexed anAddress
```
Comment (Joeysantoro) [inline contracts/pcv/IPCVGuardian.sol:14]: nit: safeAddress is better name

File: contracts/pcv/IPCVGuardian.sol [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    );
+
+    event PCVGuardianWithdrawal(
+        address indexed pcvDeposit, 
+        address indexed destination, 
+        uint amount
```
Comment (Joeysantoro) [inline contracts/pcv/IPCVGuardian.sol:20]: nit: use uint256 here and everywhere

File: contracts/pcv/IPCVGuardian.sol [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        address indexed pcvDeposit, 
+        address indexed destination,
+        address indexed token,
+        uint amount
+    );
+
+    // ---------- Read-Only API ----------
+
+    /// @notice returns true if the the provided address is a valid destination to withdraw funds to
+    /// @param anAddress the address to check
```
Comment (Joeysantoro) [inline contracts/pcv/IPCVGuardian.sol:39]: nit: pcvDeposit is better name than anAddress

File: contracts/pcv/IPCVGuardian.sol [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    function isSafeAddress(address anAddress) external view returns (bool);
+
+    /// @notice returns all safe addresses
+    function getSafeAddresses() external view returns (address[] memory);
+
+    // ---------- Governor-Only State-Changing API ----------
+
+    /// @notice governor-only method to set an address as "safe" to withdraw funds to
+    /// @param anAddress the address to set as safe
+    function setSafeAddress(address anAddress) external;
```
Comment (Joeysantoro) [inline contracts/pcv/IPCVGuardian.sol:49]: nit: pcvDeposit is better name than anAddress

File: contracts/pcv/IPCVGuardian.sol [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    // ---------- Governor-or-Guardian-Only State-Changing API ----------
+
+    /// @notice governor-or-guardian-only method to un-set an address as "safe" to withdraw funds to
+    /// @param anAddress the address to un-set as safe
+    function unsetSafeAddress(address anAddress) external;
```
Comment (Joeysantoro) [inline contracts/pcv/IPCVGuardian.sol:55]: nit: pcvDeposit is better name than anAddress

File: contracts/pcv/PCVGuardian.sol [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    /// @param safeAddress the destination address to withdraw to
+    /// @param amount the amount to withdraw
+    /// @param unpauseBefore if true, the pcv contract will be unpaused before the withdraw
+    /// @param pauseAfter if true, the pcv contract will be paused after the withdraw
+    function withdrawToSafeAddress(address pcvDeposit, address safeAddress, uint amount, bool unpauseBefore, bool pauseAfter) external override onlyGuardianOrGovernor() {
+        require(isSafeAddress(safeAddress), "Provided address is not a safe address!");
+
+        if (unpauseBefore) {
+            ICoreRef(pcvDeposit).unpause();
+        }
```
Comment (Joeysantoro) [inline contracts/pcv/PCVGuardian.sol:63]: Unpause before is always true if its currently paused, should remove param and do this dynamically

File: test/unit/pcv/PCVGuardian.test.ts [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+import { PCVGuardian__factory } from '@custom-types/contracts/factories/PCVGuardian__factory';
+import { forceEth } from '@test/integration/setup/utils';
+
+// This will theoretically make the error stack actually print!
+chai.config.includeStack = true;
+
+// Import if needed, just a helper.
+// const toBN = ethers.BigNumber.from;
+
+describe.only('PCV Guardian', function () {
```
Comment (Joeysantoro) [inline test/unit/pcv/PCVGuardian.test.ts:23]: .only

File: test/unit/pcv/PCVGuardian.test.ts [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    // add any addresses you want to impersonate here
+    const impersonatedAddresses = [userAddress, pcvControllerAddress, governorAddress, guardianAddress];
+
+    for (const address of impersonatedAddresses) {
+      await hre.network.provider.request({
+        method: 'hardhat_impersonateAccount',
+        params: [address]
+      });
+
+      impersonatedSigners[address] = await ethers.getSigner(address);
```
Comment (Joeysantoro) [inline test/unit/pcv/PCVGuardian.test.ts:56]: getImpersonatedSigner cleaner

**Review by Joeysantoro — CHANGES_REQUESTED** [review]
Very good overall, one main comment and a bunch of nits

**Comment by ditchfieldcaleb** [comment]
todo: make modifiers onlyGovernorOrAdminOrGuardian too

**Comment by ditchfieldcaleb** [comment]
> todo: make modifiers onlyGovernorOrAdminOrGuardian too

@Joeysantoro what did you mean by this?

File: contracts/pcv/PCVGuardian.sol [hunk 12]
```diff
@@ -11,54 +11,72 @@ contract PCVGuardian is IPCVGuardian, CoreRef {
...
     /// @param safeAddress the destination address to withdraw to
     /// @param amount the amount to withdraw
-    /// @param unpauseBefore if true, the pcv contract will be unpaused before the withdraw
     /// @param pauseAfter if true, the pcv contract will be paused after the withdraw
-    function withdrawToSafeAddress(address pcvDeposit, address safeAddress, uint256 amount, bool unpauseBefore, bool pauseAfter) external override onlyGuardianOrGovernor() {
+    function withdrawToSafeAddress(address pcvDeposit, address safeAddress, uint256 amount, bool pauseAfter) external override isGovernorOrGuardianOrAdmin() {
         require(isSafeAddress(safeAddress), "Provided address is not a safe address!");
 
-        if (unpauseBefore) {
+        if (CoreRef(pcvDeposit).paused()) {
```
Comment (Joeysantoro) [inline contracts/pcv/PCVGuardian.sol:79]: Nit: Prefer some kind of IPausable interface to casting as CoreRef here.

File: contracts/pcv/PCVGuardian.sol [hunk 13]
```diff
@@ -97,12 +114,11 @@ contract PCVGuardian is IPCVGuardian, CoreRef {
...
-    /// @param unpauseBefore whether to unpause the pcv before withdrawing
     /// @param pauseAfter whether to pause the pcv after withdrawing
-    function withdrawERC20ToSafeAddress(address pcvDeposit, address safeAddress, address token, uint256 amount, bool unpauseBefore, bool pauseAfter) external override onlyGuardianOrGovernor() {
+    function withdrawERC20ToSafeAddress(address pcvDeposit, address safeAddress, address token, uint256 amount, bool pauseAfter) external override isGovernorOrGuardianOrAdmin() {
         require(isSafeAddress(safeAddress), "Provided address is not a safe address!");
 
-        if (unpauseBefore) {
+        if (CoreRef(pcvDeposit).paused()) {
             ICoreRef(pcvDeposit).unpause();
         }
```
Comment (Joeysantoro) [inline contracts/pcv/PCVGuardian.sol]: This section should be an internal helper method imo

**Review by Joeysantoro — APPROVED** [review]
Two nits, LGTM

File: contracts/pcv/PCVGuardian.sol (hunk 12 above)
Comment (ditchfieldcaleb) [reply contracts/pcv/PCVGuardian.sol:79]: we can use Pausable (which is abstract anyways) for checking pause state, but pause() and unpause() themselves are defined in ICoreRef - doesn't makes sense to me to redefine them in another interface

Comment (ditchfieldcaleb) [reply contracts/pcv/PCVGuardian.sol:79]: updated with Pausable where I could

**Pull request merged**